### ISSUE TITLE
Code style changes?

### ISSUE THREAD

**Issue by iangilman** [openseadragon/openseadragon#456]
Now that I'm more actively developing on this project, I'm confronted more directly with the fact that the OpenSeadragon code style is a little unusual. I wonder if it might be worth moving towards something a little more standard. 

The list so far:
- Remove spaces inside parentheses and square brackets.
- Keeping private variables in the objects themselves (rather than hiding them via closures, including the THIS[hash] structure) but prefixing them with _. This sends a clear message that those properties aren't to be relied on, but still leaves them connected to the object for better organization and debugability. 
- Adding $.console.assert at the beginning of functions so we error when we're not given the right params.
- `self` instead of `_this`.

There are probably more things like this we could identify. Anyone else have something to add? 

If we all agree we'd like to move away from the spaces inside the parentheses, unfortunately that would be a big task, and a major discontinuity in the repo (nearly every line would change). One approach would be to just go for it and rip the Band-Aid off quick. Another approach would be to hold it as an ideal to move toward and change the style as code is touched. At the moment I'm leaning toward the latter approach.

What do you think? Anything to add?

**Comment by iangilman** [comment]
Alrighty, now that the collections branch has been merged, I suppose now is the time to do this! @msalsbery Are you still interested, or should I take a whack?

**Comment by msalsbery** [comment]
I'm down, unless you already finished!

**Comment by msalsbery** [comment]
@iangilman ^

**Comment by iangilman** [comment]
Nope...go for it :)

**Comment by edsilv** [comment]
I've been using TypeScript "in anger" for a couple of years now: 

https://github.com/UniversalViewer/universalviewer

Recent developments with Angular:

http://blogs.msdn.com/b/typescript/archive/2015/03/05/angular-2-0-built-on-typescript.aspx

and Google's "SoundScript":

https://developers.google.com/v8/experiments

would seem to assuage any doubts of its future viability.

I've grown used to the luxury of compile-time errors, but types are optional, and TS can be used simply as a way of transpiling ES6 code into idiomatic js.

To address a couple of the points above, TypeScript has private members on classes, and the compiled js uses `_this`.

Obviously porting to TypeScript doesn't really fall under the definition of a code style change :-)

I just thought I'd put this out there as I'd really love to see a future version of OSD in TypeScript and would be willing to help, even if it were just as an experimental project.

**Comment by iangilman** [comment]
@edsilv Yes, I don't think we'll be converting over to typescript anytime soon, but you're welcome to start a typescript fork if you want! 

In other news, @jedateach has pointed out another possibility for codestyle: https://github.com/flet/semistandard

**Comment by philipgiuliani** [comment]
Oh now i finally understand why you removed the spaces when you added setClip and was really confused..

I would also really like to help out with this and go for a more "normal" style :+1:

**Comment by iangilman** [comment]
Excellent. @msalsbery appears to have become distracted with other things, so feel free to take a whack at it if you'd like. :-)

**Comment by msalsbery** [comment]
:+1:

**Comment by jedateach** [comment]
May also be useful: http://editorconfig.org/

**Comment by iangilman** [comment]
@jedateach Indeed! @philipgiuliani has just added one with #637 and #640.

**Comment by avandecreme** [comment]
In the end we went more for this approach: 
>Another approach would be to hold it as an ideal to move toward and change the style as code is touched

So maybe we should add a CONTRIBUTING.md file explaining all that and add hound style checking (see #780)

**Comment by iangilman** [comment]
Sounds good. I guess @msalsbery and I keep not getting around to doing the massive shift.
